uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,1044 @@
+{
+  T new_backend_;
+  std::vector<T> * backends_;
+  size_t enter_size_;
+public:
+
+  context_helper():
+    backends_(nullptr),
+    enter_size_(size_t(-1))
+    {}
+
+  bool init(std::vector<T> & backends, T new_backend)
+    {
+      backends_ = &backends;
+      new_backend_ = std::move(new_backend);
+      return true;
+    }
+
+  bool enter()
+    {
+      enter_size_ = backends_->size();
+      try { backends_->push_back(new_backend_); }
+      catch(std::bad_alloc&)
+      {
+        PyErr_NoMemory();
+        return false;
+      }
+      return true;
+    }
+
+  bool exit()
+    {
+      bool success = (enter_size_ + 1 == backends_->size()
+                      && backends_->back() == new_backend_);
+      if (enter_size_ < backends_->size())
+        backends_->resize(enter_size_);
+
+      if (!success)
+        PyErr_SetString(PyExc_RuntimeError,
+                        "Found invalid context state while in __exit__");
+      return success;
+    }
+};
+
+
+struct SetBackendContext
+{
+  PyObject_HEAD
+
+  context_helper<backend_options> ctx_;
+
+  static void dealloc(SetBackendContext * self)
+    {
+      self->~SetBackendContext();
+      Py_TYPE(self)->tp_free(self);
+    }
+
+  static PyObject * new_(PyTypeObject * type, PyObject * args, PyObject * kwargs)
+    {
+      auto self = reinterpret_cast<SetBackendContext *>(type->tp_alloc(type, 0));
+      if (self == nullptr)
+        return nullptr;
+
+      // Placement new
+      self = new (self) SetBackendContext;
+      return reinterpret_cast<PyObject *>(self);
+    }
+
+  static int init(
+    SetBackendContext * self, PyObject * args, PyObject * kwargs)
+    {
+      static const char * kwlist[] = {"backend", "coerce", "only", nullptr};
+      PyObject * backend = nullptr;
+      PyObject * coerce = nullptr;
+      PyObject * only = nullptr;
+
+      if (!PyArg_ParseTupleAndKeywords(
+            args, kwargs,
+            "O|O!O!", (char**)kwlist,
+            &backend,
+            &PyBool_Type, &coerce,
+            &PyBool_Type, &only))
+        return -1;
+
+
+      auto domain = backend_to_domain_string(backend);
+      if (domain.empty())
+        return -1;
+      backend_options opt;
+      opt.backend = py_ref::ref(backend);
+      opt.coerce = (coerce == Py_True);
+      opt.only = (only == Py_True);
+
+      try
+      {
+        if (!self->ctx_.init(local_domain_map[domain].preferred, std::move(opt)))
+          return -1;
+      }
+      catch(std::bad_alloc&)
+      {
+        PyErr_NoMemory();
+        return -1;
+      }
+
+      return 0;
+    }
+
+  static PyObject * enter__(SetBackendContext * self, PyObject * /*args*/)
+    {
+      if (!self->ctx_.enter())
+        return nullptr;
+      Py_RETURN_NONE;
+    }
+
+  static PyObject * exit__(SetBackendContext * self, PyObject * /*args*/)
+    {
+      if (!self->ctx_.exit())
+        return nullptr;
+      Py_RETURN_NONE;
+    }
+};
+
+
+struct SkipBackendContext
+{
+  PyObject_HEAD
```

Comment:
This would be moved into the base class.

Reply:
I use `context_helper` on line 735 without creating any python objects.

Reply:
Fair enough, still would be nice to have a base class.